---
title: "Low_High_Mut_R"
output: html_document
date: "2022-10-31"
---
#uploading data and defining a column 
```{r}
SeqData <- read.table('//Users/nagendra//Desktop//Bioinformatics//Project//Data//ACTG5257_RT_fasta.txt',header=FALSE,fill=TRUE) 
WeekNumber <- SeqData$V1

```

#deleting rows with intermediates between 0 and final dat
```{r}
SeqData<- SeqData[-c(21,22,61,62,225,226,337,338,375,376,401,402,403,404,617,618,707,708,909,910,985,986,571,572,591,592,799,800,871,872,889,890,893,929,947,973,997,1011,1061,1067,1077,1079,487,488),]

```

#deleting rows that only have day 0 but not a second day
```{r}
for(each in 1:length(WeekNumber)){
  if (SeqData$V8[each] == 0 && SeqData$V8[each] == SeqData$V8[each+2] && is.na(SeqData$V8[each])==FALSE){
    SeqData <- SeqData[-c(each,each+1),]
  } else{
    next
  } 
}
```


#Taking out patients with sequence between 0 and final(in progress)
```{r}
for(t in 1:length(SeqData$V2)){
  if (SeqData$V5[t] == SeqData$V5[t+2] && SeqData$V5[t] == SeqData$V5[t+4] && is.na(SeqData$V5[t])==FALSE){
    SeqData <- SeqData[-c(t+2,t+3),]
  } else{
    next
  } 
}
```

#splitting into week0 and end week sequences
```{r}
Week0Seq <- c()
EndWeekSeq <- c()
for( x in 1:length(WeekNumber)){
  if (SeqData$V8[x] == 0 && is.na(SeqData$V8[x])==FALSE){
   Week0Seq <- append(Week0Seq,SeqData$V1[x+1]) 
 }else if(is.na(SeqData$V8[x])){
   next
 } else if (is.numeric(SeqData$V8[x])){
    EndWeekSeq<- append(EndWeekSeq,SeqData$V1[x+1])
  } else(
    next
  )
}
  
```

#creating dataframe for week data
```{r}
StartVsEndDF <- data.frame(Week0Seq, EndWeekSeq)
```

#stringsplit
```{r}
VectEnd<- unlist(strsplit(EndWeekSeq, ","))
Vect0<- unlist(strsplit(Week0Seq, ","))
```

#removing periods in sequences
```{r}
StartVsEndDF$Week0Seq <- gsub("\\.","", StartVsEndDF$Week0Seq)
StartVsEndDF$EndWeekSeq <- gsub("\\.","", StartVsEndDF$EndWeekSeq)
```
https://blast.ncbi.nlm.nih.gov/Blast.cgi?BLAST_SPEC=blast2seq&LINK_LOC=align2seq&PAGE_TYPE=BlastSearch 

#adding patient ID next to sequence (in progress)
```{r}
PatientID <- c()
for(mama in 1:length(SeqData$V1)){
  if(SeqData$V2[mama] == SeqData$V2[mama+2] && is.na(SeqData$V2[mama])==FALSE){
   PatientID <- append(PatientID, SeqData$V2[mama])
  }else if(is.na(SeqData$V2[mama])){
    next
  } else(next)
  }
 
  
```
INDEXING second solution ***

```{r}
idx <- c()
#trying to move sequences for day0 and end day for each patient to a different dataframe
#Loop comments
#making sure the patient names are the same, the day for the row is 0 
#making sure the indexed box is not empty or NA
for(h in 1:length(SeqData$V1)){
  if(SeqData$V5[h]==SeqData$V5[h+2] && SeqData$V8[h] == 0 && is.na(SeqData$V8[h])==FALSE && SeqData$V5[h]!= "" && SeqData$V8[h]!=SeqData$V8[h+2]){
    #if all those conditions are met, the row number is indexed and saved in idx
    idx <- append(idx, h)
  }
}
  
StartSeq<- SeqData$V1[idx+1]
EndSeq <- SeqData$V1[idx+3]
PTID <- SeqData$V5[idx]
StartDayNumber <- SeqData$V8[idx]
EndDayNumber <- SeqData$V8[idx+2]

```

Making a dataframe with the saved rows
```{r}
CompleteDF <- data.frame(StartSeq, EndSeq, PTID, StartDayNumber, EndDayNumber)
```

Removing periods before nucleotide sequences
```{r}
CompleteDF$StartSeq <- gsub("\\.","", CompleteDF$StartSeq)
CompleteDF$EndSeq <- gsub("\\.","", CompleteDF$EndSeq)
```

Saving dataframe to a text file
```{r}

write.table(CompleteDF,'/Users/nagendra/Desktop/Bioinformatics/Project/Data/DataForBLAST.csv', sep = ",")
```

Updating blast sequence data
```{r}
BLAST <- read.table('/Users/nagendra/Desktop/Bioinformatics/Project/Data/BLASTFinalAlignment.csv', sep= ",")
```
Getting Blast sequences cleaned up 
```{r}
ydx<-c()
for(z in 1:length(BLAST$V1)){
  if(BLAST$V1[z]==BLAST$V2[z]){
   ydx <- append(ydx,z) 
  }
}

PatientName <- BLAST$V1[ydx]
SimilarityPct <- BLAST$V3[ydx]
BPlength <- BLAST$V4[ydx]

```

Making comparison BLAST dataframe
```{r}
BLASTCompare <- data.frame(PatientName, SimilarityPct, BPlength, EndDayNumber)
```

Trouble shooting
```{r}
numYes <- c(0)
numNo <- c(0)
for(s in 1:242){
  if(BLASTCompare$PatientName[s] == CompleteDF$PTID[s]){
    numYes <- numYes + 1
  }else{
    numNo <- numNo + 1
  }
}
```

